OpenEdX/Managers/PushNotificationsManager/PushNotificationsManager.swift

```diff
@@ -94,4 +102,58 @@ class PushNotificationsManager {
             listener.didReceiveRemoteNotification(userInfo: userInfo)
         }
     }
+    
+    func synchronizeToken() {
+        for provider in providers {
+            provider.synchronizeToken()
+        }
+    }
+    
+    func refreshToken() {
+        for provider in providers {
+            provider.refreshToken()
+        }
+    }
+}
+
+// MARK: - MessagingDelegate
+extension PushNotificationsManager: MessagingDelegate {
+    func messaging(_ messaging: Messaging, didReceiveRegistrationToken fcmToken: String?) {
+        for provider in providers where provider is MessagingDelegate {
+            (provider as? MessagingDelegate)?.messaging?(messaging, didReceiveRegistrationToken: fcmToken)
+        }
+    }
+}
+
+// MARK: - UNUserNotificationCenterDelegate
+extension PushNotificationsManager: UNUserNotificationCenterDelegate {
+    @MainActor
+    func userNotificationCenter(
+        _ center: UNUserNotificationCenter,
+        willPresent notification: UNNotification
+    ) async -> UNNotificationPresentationOptions {
+        // With swizzling disabled you must let Messaging know about the message, for Analytics
+        Messaging.messaging().appDidReceiveMessage(notification.request.content.userInfo)
+        
+        // Show alert if application is active
+        if let pushManager = Container.shared.resolve(PushNotificationsManager.self),
+           UIApplication.shared.applicationState == .active {
+            pushManager.didReceiveRemoteNotification(userInfo: notification.request.content.userInfo)
+        }
+        
+        return [[.list, .banner, .sound]]
+    }
+    
+    @MainActor
+    func userNotificationCenter(
+        _ center: UNUserNotificationCenter,
+        didReceive response: UNNotificationResponse
+    ) async {
+        let userInfo = response.notification.request.content.userInfo
+        
+        // With swizzling disabled you must let Messaging know about the message, for Analytics
+        Messaging.messaging().appDidReceiveMessage(userInfo)
```

Comment:
I guess the right place for this is `FCMListener`.

Reply:
It requires extra abstraction for PushNotificationsProvider; it can create overhead.
For me, it seems ok to leave just 1 line here.
wdyt?

Reply:
It won't required any abstraction as the abstraction is already there. The `FCMListener: PushNotificationsListener` is implementing the `PushNotificationsListener` protocol, you just need to implement optional method `didReceiveRemoteNotification` and thats it.

Reply:
@saeedbashir, but if we override didReceiveRemoteNotification in FCMListener, the extension with default implementation (navigation) will stop working 🤷‍♂️.
<img width="665" alt="image" src="https://github.com/openedx/openedx-app-ios/assets/127732735/02f30201-021a-42fe-a9a4-a35e1ebb6c8c">

Reply:
We need one more function in `PushNotificationsListener` for this or refactor `didReceiveRemoteNotification` and not use it for navigation.

Reply:
Strange, I guess it got missed while writing the abstraction layer.  Ideally, every `PushNotificationsListener` should implement this piece of code. So first, the listener will check whether the push is for that listener (Braze, FCM etc) or not, and then do the required work before starting the navigation. 

The other reason for giving each listener its own implementation is that maybe data will be received in a different format (hierarchy), and a parsing may be required to generate a PushLink from the data.

Reply:
Sure, I'll update it now, I think it makes sense

Reply:
@saeedbashir I've updated implementation of the `PushNotificationsManager`, please check.